### PR TITLE
bug in set constrained

### PR DESCRIPTION
**Pull request by nicola-giuliani**

### CODE REVIEW

**Comment by nicola-giuliani** [comment]
I believe to have found a bug in set_constrained_dofs_to_zero.

I have written the test that should fail without the modification I have written. @ESeNonFossiIo if this is the case with PETSc should be a blood-bath.

File: source/pidomus.cc [hunk 1]
```diff
@@ -1200,11 +1200,20 @@ template <int dim, int spacedim, typename LAC>
...
+  if (global_partitioning.n_elements() > 0)
     {
-      auto j = global_partitioning.nth_index_in_set(i);
-      if (constraints.is_constrained(j))
-        v[j] = 0;
+      auto k = global_partitioning.nth_index_in_set(0);
+      if (constraints.is_constrained(k))
+        v[k] = 0;
+      else
+        v[k] = v[k];
```
Comment (ESeNonFossiIo) [inline source/pidomus.cc:1209]: Is this line necessary?

File: source/pidomus.cc [hunk 2]
```diff
@@ -1200,11 +1200,20 @@ template <int dim, int spacedim, typename LAC>
...
+      for (unsigned int i = 1; i < global_partitioning.n_elements(); ++i)
+        {
+          auto j = global_partitioning.nth_index_in_set(i);
+          if (constraints.is_constrained(j))
+            v[j] = 0;
+        }
+      v.compress(VectorOperation::insert);
```
Comment (ESeNonFossiIo) [inline source/pidomus.cc:1216]: This line is the only line missing in the previous version. Isn't it so?

File: source/pidomus.cc (hunk 1 above)
Comment (luca-heltai) [inline source/pidomus.cc:1209]: It think it is a good idea. If I understand correctly, this ensures that at least once (on the first element), each processor does one set operation.

File: source/pidomus.cc (hunk 2 above)
Comment (luca-heltai) [inline source/pidomus.cc:1216]: This is fine, and with this probably we shouldn't need the hack above. @nicola-giuliani , did you try without the hack above first, and it didn't work, or was the hack above necessary if one did not have the compress here?

File: source/pidomus.cc (hunk 1 above)
Comment (ESeNonFossiIo) [inline source/pidomus.cc:1209]: Can you explain to me why we should take care of it?
I do not understand why each processor has to do at least one set operation.

**Pull request merged**